### PR TITLE
Issue 377

### PR DESCRIPTION
**Pull request by alvaro-castillo**

### CODE REVIEW

File: skfda/_utils/_utils.py [hunk 1]
```diff
@@ -730,6 +730,21 @@ def _classifier_fit_depth_methods(
     return classes, class_depth_methods_
 
 
+def _fit_feature_transformer(
```
Comment (vnmabus) [inline skfda/_utils/_utils.py:733]: Where is this function used?

File: skfda/preprocessing/dim_reduction/feature_extraction/_fda_feature_union.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+    """
+
+    def __init__(
+        self,
+        transformer_list,
+        *,
+        n_jobs=None,
+        transformer_weights=None,
+        verbose=False,
+        np_array_output=False,
```
Comment (vnmabus) [inline skfda/preprocessing/dim_reduction/feature_extraction/_fda_feature_union.py:75]: Again, I prefer just `array_output`.

File: skfda/preprocessing/dim_reduction/feature_extraction/_fda_feature_union.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+            for i in Xs:
+                if isinstance(i, FDataGrid or FDataBasis):
+                    raise TypeError(
+                        "There are transformed instances of FDataGrid or "
+                        "FDataBasis that can't be concatenated on a NumPy "
+                        "array.",
+                    )
+            return super()._hstack(Xs)
+
+        if not isinstance(Xs[0], FDataGrid or FDataBasis):
```
Comment (vnmabus) [inline skfda/preprocessing/dim_reduction/feature_extraction/_fda_feature_union.py:97]: Again, this should also work with multivariate data.

File: skfda/preprocessing/dim_reduction/feature_extraction/_fda_feature_union.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+            raise TypeError(
+                "Transformed instance is not of type FDataGrid or"
+                " FDataBasis. It is " + type(Xs[0]),
+            )
+
+        frames = [DataFrame({Xs[0].dataset_name.lower(): Xs[0]})]
+
+        for j in Xs[1:]:
```
Comment (vnmabus) [inline skfda/preprocessing/dim_reduction/feature_extraction/_fda_feature_union.py:105]: You can create the `DataFrame` without looping.

**Review by vnmabus — CHANGES_REQUESTED** [review]

File: skfda/_utils/_utils.py (hunk 1 above)
Comment (alvaro-castillo) [reply skfda/_utils/_utils.py:733]: That function is used on issue 376, I forgot to delete it from this branch.

**Review by alvaro-castillo — COMMENTED** [review]
I have fixed what you remarked

File: skfda/preprocessing/dim_reduction/feature_extraction/_fda_feature_union.py [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+
+    Then we need to import the transformers we want to use
+    >>> from skfda.preprocessing.dim_reduction.feature_extraction import FPCA
+    >>> from skfda.representation import EvaluationTransformer
+
+    Finally we import the union and apply fit and transform
+    >>> from skfda.preprocessing.dim_reduction.feature_extraction
+    ... import FdaFeatureUnion
+    >>> union = FdaFeatureUnion([
+    ...    ("Eval", EvaluationTransformer()),
```
Comment (vnmabus) [inline skfda/preprocessing/dim_reduction/feature_extraction/_fda_feature_union.py:63]:
```suggestion
    ...    ("eval", EvaluationTransformer()),
```

File: skfda/preprocessing/dim_reduction/feature_extraction/_fda_feature_union.py [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+    ...    ("fpca", FPCA()), ], array_output=True)
+    >>> union.fit_transform(X)
```
Comment (vnmabus) [inline skfda/preprocessing/dim_reduction/feature_extraction/_fda_feature_union.py:65]: This test is not passing. Also an example where the new features are concatenated with the original curve would be good.

File: tests/test_fda_feature_union.py [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+    """Check the Fda Feature Union module."""
+
+    def setUp(self) -> None:
+        """Fetch the Berkeley Growth Study dataset."""
+        self.X = fetch_growth(return_X_y=True)[0]
+
+    def test_incompatible_array_output(self) -> None:
+        """Check that the transformer returns a ndarray."""
+        u = FdaFeatureUnion(
+            [("EvaluationT", EvaluationTransformer(None)), ("fpca", FPCA())],
```
Comment (vnmabus) [inline tests/test_fda_feature_union.py:30]:
```suggestion
            [("eval", EvaluationTransformer(None)), ("fpca", FPCA())],
```

**Review by vnmabus — CHANGES_REQUESTED** [review]

File: skfda/preprocessing/dim_reduction/feature_extraction/_fda_feature_union.py (hunk 6 above)
Comment (alvaro-castillo) [reply skfda/preprocessing/dim_reduction/feature_extraction/_fda_feature_union.py:65]: I am wondering how to concatenate the original data to the transformed one as both  have different sizes.
In addition, X.data_matrix is an array of functions, and  each of them an array of array points (dimension 3). The transformed data instead is an array of arrays containing transformed points (dimension 2).

Comment (vnmabus) [reply skfda/preprocessing/dim_reduction/feature_extraction/_fda_feature_union.py:65]: You are thinking this wrong. The transformed data always has the same number of samples as the original data. The concatenation is done as separate columns, and all of them have the same number of rows.

Comment (alvaro-castillo) [reply skfda/preprocessing/dim_reduction/feature_extraction/_fda_feature_union.py:65]: Okay, I have developed an algorithm that is able to concatenate them, but it is too complex for writing it into the example (it includes two nested loops), should I create it as a function on the module FeatureUnion and call it?

**Review by alvaro-castillo — COMMENTED** [review]
I have fixed everything

File: skfda/preprocessing/dim_reduction/feature_extraction/_fda_feature_union.py [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+    Finally we apply fit and transform
+    >>> union = FdaFeatureUnion(
+    ...     [
+    ...        ("eval", EvaluationTransformer()),
+    ...        ("fpca", FPCA()),
+    ...     ],
+    ...     array_output=True,
+    ... )
+    >>> transformed_data = union.fit_transform(X)
+    >>> transformed_data
```
Comment (vnmabus) [inline skfda/preprocessing/dim_reduction/feature_extraction/_fda_feature_union.py:71]: Maybe it is better to round the values, for clarity and reproducibility.

File: skfda/preprocessing/dim_reduction/feature_extraction/_fda_feature_union.py [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+           [ 68.6       ,  73.6       ,  78.6       , ..., -19.49404628,
+             12.76825883,   0.70188222],
+           [ 79.9       ,  82.6       ,  84.8       , ...,  19.28399897,
+             31.49601648,   6.54012077],
+           [ 76.1       ,  78.4       ,  82.3       , ...,  17.71973789,
+             27.7332045 ,  -1.70532625]]
+
+    We can also concatenate the result with the
+    original data on a Pandas DataFrame.
+    >>> from pandas.core.frame import DataFrame
```
Comment (vnmabus) [inline skfda/preprocessing/dim_reduction/feature_extraction/_fda_feature_union.py:88]: Don't concatenate it manually! Show an example using `FdaFeatureUnion`.

File: skfda/preprocessing/dim_reduction/feature_extraction/_fda_feature_union.py [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+    ) -> None:
+        self.np_array_output = np_array_output
+        super().__init__(
+            transformer_list,
+            n_jobs=n_jobs,
+            transformer_weights=transformer_weights,
+            verbose=verbose,
+        )
+
+    def _hstack(self, Xs) -> Union[DataFrame, ndarray, Any]:
```
[the remark this answers is not in the document]
Comment (vnmabus) [reply skfda/preprocessing/dim_reduction/feature_extraction/_fda_feature_union.py:115]: Why is `Any` a return option?

**Review by vnmabus — CHANGES_REQUESTED** [review]

File: skfda/preprocessing/dim_reduction/feature_extraction/_fda_feature_union.py (hunk 8 above)
Comment (alvaro-castillo) [reply skfda/preprocessing/dim_reduction/feature_extraction/_fda_feature_union.py:71]: You are right, I round the values of the example to two decimal values

File: skfda/preprocessing/dim_reduction/feature_extraction/_fda_feature_union.py (hunk 9 above)
Comment (alvaro-castillo) [reply skfda/preprocessing/dim_reduction/feature_extraction/_fda_feature_union.py:88]: I just noticed that EvaluationTransformer() actually gives the original data, and so we do not need to concatenate it again. I will add to the example another transformer like SRSF()

**Review by alvaro-castillo — COMMENTED** [review]
I have changed a bit the example

File: skfda/preprocessing/dim_reduction/feature_extraction/_fda_feature_union.py [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+
+from numpy import ndarray
+from pandas import DataFrame, concat
+from sklearn.pipeline import FeatureUnion
+
+from ....representation.basis import FDataBasis
+from ....representation.grid import FDataGrid
+
+
+class FdaFeatureUnion(FeatureUnion):
```
[the remark this answers is not in the document]
Comment (vnmabus) [reply skfda/preprocessing/dim_reduction/feature_extraction/_fda_feature_union.py:14]: Ignore this error with a comment.

File: skfda/preprocessing/dim_reduction/feature_extraction/_fda_feature_union.py [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+
+    Finally we apply fit and transform.
+    >>> union = FdaFeatureUnion(
+    ...     [
+    ...        ("mrmr", MinimumRedundancyMaximumRelevance()),
+    ...        ("eval", EvaluationTransformer()),
+    ...     ],
+    ...     array_output=True,
+    ... )
+    >>> np.around(union.fit_transform(X,y), decimals = 2)
```
Comment (vnmabus) [inline skfda/preprocessing/dim_reduction/feature_extraction/_fda_feature_union.py:76]: You shouldn't use whitespace around keyword parameters equals:
```suggestion
    >>> np.around(union.fit_transform(X,y), decimals=2)
```

File: skfda/preprocessing/dim_reduction/feature_extraction/_fda_feature_union.py [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+            n_jobs=n_jobs,
+            transformer_weights=transformer_weights,
+            verbose=verbose,
+        )
+
+    def _hstack(self, Xs) -> Union[DataFrame, ndarray]:
+
+        if self.array_output:
+            for i in Xs:
+                if isinstance(i, (FDataGrid, FDataBasis)):
```
Comment (vnmabus) [inline skfda/preprocessing/dim_reduction/feature_extraction/_fda_feature_union.py:107]: Why not the superclass instead?
```suggestion
                if isinstance(i, FData):
```

**Review by vnmabus — CHANGES_REQUESTED** [review]

File: skfda/preprocessing/dim_reduction/feature_extraction/_fda_feature_union.py [hunk 14]
```diff
@@ -0,0 +1,114 @@
...
+from typing import Union
+
+from numpy import ndarray
+from pandas import DataFrame
+from sklearn.pipeline import FeatureUnion
+
+from ....representation import FData
+
+
+class FdaFeatureUnion(FeatureUnion):  # type: ignore
```
Comment (vnmabus) [inline skfda/preprocessing/dim_reduction/feature_extraction/_fda_feature_union.py:13]: Shouldn't the name be `FDAFeatureUnion` instead? FDA is an acronym.

Comment (alvaro-castillo) [reply skfda/preprocessing/dim_reduction/feature_extraction/_fda_feature_union.py:13]: You are right, I didn't notice it

**Review by vnmabus — APPROVED** [review]

**Review by vnmabus — CHANGES_REQUESTED** [review]
Tests are still failing... Maybe choose a different example that it is more precise?

**Comment by alvaro-castillo** [comment]
I have changed the transformer used in the example and it seems to work

**Review by vnmabus — APPROVED** [review]

**Pull request merged**